libraries/Microsoft.Bot.StreamingExtensions/ProtocolAdapter.cs

```diff
@@ -0,0 +1,85 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.StreamingExtensions.Payloads;
+using Microsoft.Bot.StreamingExtensions.PayloadTransport;
+using Microsoft.Bot.StreamingExtensions.Utilities;
+
+namespace Microsoft.Bot.StreamingExtensions
+{
+    internal class ProtocolAdapter
+    {
+        private readonly RequestHandler _requestHandler;
+        private readonly object _handlerContext;
+        private readonly IPayloadSender _payloadSender;
+        private readonly IPayloadReceiver _payloadReceiver;
+        private readonly IRequestManager _requestManager;
+        private readonly SendOperations _sendOperations;
+        private readonly IStreamManager _streamManager;
+        private readonly PayloadAssemblerManager _assemblerManager;
+
+        public ProtocolAdapter(RequestHandler requestHandler, IRequestManager requestManager, IPayloadSender payloadSender, IPayloadReceiver payloadReceiver, object handlerContext = null)
+        {
+            _requestHandler = requestHandler;
+            _handlerContext = handlerContext;
+            _requestManager = requestManager;
+            _payloadSender = payloadSender;
+            _payloadReceiver = payloadReceiver;
+
+            _sendOperations = new SendOperations(_payloadSender);
+            _streamManager = new StreamManager(OnCancelStream);
+            _assemblerManager = new PayloadAssemblerManager(_streamManager, OnReceiveRequest, OnReceiveResponse);
+
+            _payloadReceiver.Subscribe(_assemblerManager.GetPayloadStream, _assemblerManager.OnReceive);
+        }
+
+        public async Task<ReceiveResponse> SendRequestAsync(Request request, CancellationToken cancellationToken)
+        {
+            var requestId = Guid.NewGuid();
+
+            // wait for the response befores sending the request , so that we dont hit the case
+            // where response is received before request is sent. This would lead to no tcs
```

Comment:
This comment doesn't match the code below. The Send is happening immediatly (it's async).